File: platform/polyglot/rag_embedding_client/py/rag_embedding_client/exceptions.py
"""Custom exceptions for the RAG embedding client.

``EmbeddingBatchRejectedError`` is raised when one or more sub-batches are
rejected with HTTP 403 Forbidden during ``embed_documents()``.  It carries
any partial embeddings that succeeded and a list of
``ForbiddenRejection`` records describing the rejected sub-batches.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class ForbiddenRejection:
    """Details about a single sub-batch rejected with HTTP 403."""

    batch_index: int
    """0-based index of the sub-batch within the overall embed_documents() call."""

    text_count: int
    """Number of texts in the rejected sub-batch."""

    content_preview: str
    """First 200 characters of the first text in the sub-batch."""

    status_code: int = 403
    """HTTP status code returned by the upstream server."""

    error_body: str = ""
    """Raw response body (truncated to 1000 chars)."""


@dataclass
class EmbeddingBatchRejectedError(Exception):
    """Raised when sub-batches are rejected with HTTP 403 Forbidden.

    Callers can inspect ``partial_embeddings`` for the results that
    succeeded and ``rejections`` for details about the rejected
    sub-batches.
    """

    message: str
    partial_embeddings: list[list[float]] = field(default_factory=list)
    rejections: list[ForbiddenRejection] = field(default_factory=list)

    def __str__(self) -> str:
        n_ok = len(self.partial_embeddings)
        n_rejected = len(self.rejections)
        return f"{self.message} (partial={n_ok}, rejected={n_rejected})"

File: platform/polyglot/rag_embedding_client/mjs/src/exceptions.mjs
/**
 * @fileoverview Custom exceptions for the RAG embedding client.
 *
 * {@link EmbeddingBatchRejectedError} is thrown when one or more sub-batches
 * receive HTTP 403 Forbidden during `embedDocuments()`.  It carries any
 * partial embeddings that succeeded and a list of {@link ForbiddenRejection}
 * records describing the rejected sub-batches.
 */

/**
 * Details about a single sub-batch rejected with HTTP 403.
 */
export class ForbiddenRejection {
  /**
   * @param {object} opts
   * @param {number} opts.batchIndex - 0-based index of the sub-batch
   * @param {number} opts.textCount  - Number of texts in the rejected sub-batch
   * @param {string} opts.contentPreview - First 200 chars of the first text
   * @param {number} [opts.statusCode=403] - HTTP status code
   * @param {string} [opts.errorBody=''] - Raw response body (truncated to 1000 chars)
   */
  constructor({ batchIndex, textCount, contentPreview, statusCode = 403, errorBody = '' }) {
    this.batchIndex = batchIndex;
    this.textCount = textCount;
    this.contentPreview = contentPreview;
    this.statusCode = statusCode;
    this.errorBody = errorBody;
  }
}

/**
 * Thrown when sub-batches are rejected with HTTP 403 Forbidden.
 *
 * Callers can inspect `partialEmbeddings` for the results that succeeded
 * and `rejections` for details about the rejected sub-batches.
 */
export class EmbeddingBatchRejectedError extends Error {
  /**
   * @param {string} message
   * @param {object} [opts]
   * @param {number[][]} [opts.partialEmbeddings=[]]
   * @param {ForbiddenRejection[]} [opts.rejections=[]]
   */
  constructor(message, { partialEmbeddings = [], rejections = [] } = {}) {
    super(message);
    this.name = 'EmbeddingBatchRejectedError';
    this.partialEmbeddings = partialEmbeddings;
    this.rejections = rejections;
  }
}
